### DISCUSSION TITLE
OKLab color picker: Why?

### DISCUSSION THREAD

**Discussion by Semphriss** [SuperTux/supertux#2323] [Q&A]
In #1816, a new method of choosing colors was added to replace the traditional RGB one.

My initial though is that OkLab would be too technical for SuperTux, and that the advantages it is meant to bring (such as lightness control) are too "scientific" and do not fit within SuperTux's surreal art style. I thought the upsides of OkLab would come up as people use it, but I've never come to observe any. In fact, after asking around for opinions, it would seem that most users find it difficult to choose the colors they want with OkLab and keep using the traditional RGB selector.

I am usually glad to see improvements and new techniques added to SuperTux. But in this case, to the maximum extent of my knowledge, the new color picker does not allow for creating new color combinations, nor does it make choosing colors easier, nor does it allow creating any special color palette that would fit the game's style, nor does it offer any other advantage over RGB that is meaningful in the context of SuperTux.

My question is: **Why was this change initially believed to be needed, useful, or relevant?** Perhaps we can try to find out if we can change something to get the most out of the new color picker, or if it really doesn't fit in SuperTux, we may remove it to keep things simple for the users.

(@HybridDog if you're here, since you're the author of the PR, your opinion would be most useful)

**Comment by HybridDog** [comment]
I was inspired by the explanations in [these blog posts](https://bottosson.github.io/).

Initially SuperTux had a sRGB colour picker with sliders. A sRGB colour picker is problematic in my opinion because halving all three values in sRGB does neither halve the brightness physically (luminance) nor perceptually (lightness). Therefore I've added numbers which show the linear brightness in percent. 
I think that in addition to RGB colours, a user may want to be able to easily change the hue, chroma and lightness, too, so I've added a perceptual colour picker.

It is possible to implement a 2D instead of 1D OKLab colour picker, see for example https://bottosson.github.io/misc/colorpicker/#00c779.
I've only implemented a simple one based on my own ideas, and a 2D picker could be much more useful for regular players.

**Reply by Semphriss** [reply]
> [...] a user may want to be able to easily change the hue, chroma and lightness [...]

This is the part I'm not sure I understand - why would a user need control over lightness across colors? What are the use cases, and how common are they?

I always presumed that in the rare case there would be an actual use for it, someone could simply use a picker online and then translate that into RGB manually. This shouldn't be a significantly slowing factor, unless the task would have be repeated all the time, not only in certain specific cases - and I'm not sure why it would have to happen all the time.

Also, including it in the game directly makes it feel like it should be the "proper way" to do it, but I'm not sure if it should indeed be the proper way to do it, and the reasons why. Most of the game works in traditional RGB (I'm thinking most notably about the lanterns), and it feels like the cartoony art style fits much better with regular RGB colors than a physically correct color space. I see the use case for proper color management in apps that need professional control like Blender, but I don't see the advantages of physical correctness in a game like SuperTux.

I'm looking mostly for why SuperTux needs OkLab, and how its advantages apply in the specific case of SuperTux.

**Reply by HybridDog** [reply]
Blue (r=0, g=0, b=1) looks significantly darker to humans than e.g. green (r=0, g=1, b=0). With the hue slider, a user can change the colour from e.g. red to blue while preserving the lightness (perceived brightness). With RGB sliders, this is much more difficult (at least for me).

With the lightness slider, the perceived brightness can be changed while the perceived hue and chroma stay the same. This would not be possible with a luminance slider, for example because a physical mix of white and dark blue light looks violet to humans, as can be seen [on those example gradients](https://bottosson.github.io/posts/colorwrong/#comparisons).

There is enough space in the colour picker window for additional OKLab sliders; without them, this space would be simply unused. My biggest concern when implementing the OKLab sliders was maintainability of the code, so I've moved most of the OKLab implementation to a separate file.

The OKLab colourspace is a perceptual and not a physical one. Using linearized RGB for rendering is [another issue](https://github.com/SuperTux/supertux/issues/1799).

There are other possible uses of OKLab in SuperTux than a colour picker, for example, perceptual gradients as level backgrounds. In my opinion, the non-linear OKLab is better suited for a cartoony art style than the non-linear sRGB.

**Reply by Semphriss** [reply]
I'm not sure we understand each other there.

I'm not asking for why the creator of OkLab originally decided to made it, I'm asking:

> why [does] SuperTux needs OkLab?

> why would a user need control over lightness across colors [in the first place]?

AKA: why SuperTux can't do just fine without OkLab?

If it can clarify my question:
RGB does not preserve perceived lightness across hues, as you said with green 0000FF vs blue 00FF00. So what? What would change if we had proper perceived lightness? Why would an end user care about it? We've passed the last 20 years with RGB, why can't we continue that way? What do we lose by not having an OkLab selector - yes, we lose easy control over perceived lightness, but what do we lose by not having control over perceived lightness? What does preceptually-correct stuff have over stuff that isn't? Why should we strive for perceptual correctness? I presume perceptual correctness is not an end goal in and of itself, but rather, that you are using it to achieve a certain effect or something; what is that "something"?

**Reply by HybridDog** [reply]
A hue, chroma and lightness selection in a colour picker is convenient for the user in my opinion, but it is not essential and having a feature-rich colour picker is a very low priority in SuperTux. I don't know how other people perceive the usefulness of the OKLab sliders. You can remove them or add a setting to disable them if you think this is needed. 

> AKA: why SuperTux can't do just fine without OkLab?

SuperTux can do fine without OkLab. 

> What would change if we had proper perceived lightness?

I think with the OKLab sliders, it is more convenient to select a different colour which looks as bright as the current one. Users can also do this with the RGB sliders, but this would require some trial and error and may lead to different results depending on the physical environment.

> Why would an end user care about it?

For an end user, it should provide more convenience.

> We've passed the last 20 years with RGB, why can't we continue that way?

A perceptual colour space is not essential for SuperTux, so we can continue with only RGB sliders if required.

> What do we lose by not having an OkLab selector - yes, we lose easy control over perceived lightness, but what do we lose by not having control over perceived lightness?

We wouldn't lose much. I've implemented the OKLab selector out of personal interest and as an attempt to make colour selection in SuperTux more convenient for me. 

> What does preceptually-correct stuff have over stuff that isn't?

There are many examples (see my previous comment). Often HSV is used for a selection of hue, chroma and lightness, but in comparison to OKLab, the HSV values do not correspond well to perception.

> Why should we strive for perceptual correctness?

If the user wants to, for example, change only the hue of a background colour while retaining the perceived brightness, he/she can do this easily with an OKLab slider. The user can also do this with RGB sliders and trial and error, but depending on the phsyical environment (screen colour, colour of the lamp in the room, etc.), the resulting change in colour may be perceived different by other users in other environments. 

> I presume perceptual correctness is not an end goal in and of itself, but rather, that you are using it to achieve a certain effect or something; what is that "something"?

I use the perceptual correctness to achieve a more accurate hue, chroma and lightness selection than those of HSV.

**Reply by Semphriss** [reply]
> I've implemented the OKLab selector out of personal interest and as an attempt to make colour selection in SuperTux more convenient for me.

This is the answer I was looking for.

I wasn't sure if you added it because having consistent lightness would, for example, allow users to make their levels more beautiful by default because _\<some paper or Wikipedia article that explains why perceptually-correct images are more beautiful than images that aren't\>_.

In a way, one could say I was looking for the equivalent of [the argument for the Filmic color space in Blender](https://www.youtube.com/watch?v=m9AT7H4GGrA), but with OkLab and the concept of perception of lightness in SuperTux.

Normally, I wouldn't recommend removing it, but - the reason I started this whole thread - many people have reported that they're confused by the OkLab selector, and that they prefer using the classic RGB thing. To be fair, the concept behind OkLab is pretty advanced, and it's not really relevant in SuperTux. Since it doesn't seem to be used, I'd prefer to avoid keeping a scary feature for the users and an extra load of maintenance for the developers. As such, I would recommend, if possible, using external tools instead, such as [this website](https://products.aspose.app/svg/color-converter/oklab-to-rgb) which offers a convenient OkLab-to-RGB converter. If would prefer to retain, there's also the option of maintaining local patches (I've done that with a side feature for a while).

**Reply by HybridDog** [reply]
I guess 1D sliders are too complicated for people who don't understand the colour space, so I'm not against their removal.


> As such, I would recommend, if possible, using external tools instead, such as this website which offers a convenient OkLab-to-RGB converter.

Using external tools is inconvenient because the SuperTux colour picker doesn't yet have a field to copy or paste the current colour.

#### 2D colour selection (idea)

Here's an idea for an optional alternative extension to the colour picker.

Perhaps a simple 2D colour picker which only has fully-saturated colours would be helpful for players.
I assume that in SuperTux users are often only interested in fully-saturated colours and not greyish ones.
In addition to the current RGB sliders, the colour picker could show [this image](https://user-images.githubusercontent.com/3192173/198879594-df2ca87d-c64f-4f27-92df-61a68fd4c3b7.png) and the player can select any fully-saturated colour by clicking somewhere on the image.
Furthermore, it is possible to show the location of the current colour in the image. For reference, here are the locations of the "tango" colours:
![tango_colours](https://user-images.githubusercontent.com/3192173/198880000-cbbbeed8-837a-417e-9fa1-21ae70923f0c.svg)

**Reply by HybridDog** [reply]
> 2D colour selection (idea)

I have implemented this idea at https://github.com/SuperTux/supertux/pull/2895. Sorry for the long delay.
